### PR TITLE
Gateway Changes

### PR DESCRIPTION
**Pull request by mappum**
This PR adds seeking, IPNS path support, and root redirects to the HTTP gateway.

### CODE REVIEW

File: core/corehttp/gateway_handler.go [hunk 1]
```diff
@@ -63,8 +65,23 @@ func (i *gatewayHandler) loadTemplate() error {
 	return nil
 }
 
-func (i *gatewayHandler) ResolvePath(path string) (*dag.Node, error) {
-	return i.node.Resolver.ResolvePath(path)
+func (i *gatewayHandler) ResolvePath(p string) (*dag.Node, string, error) {
+	if strings.HasPrefix(p, "/ipns/") {
+		var err error
+		p, err = i.node.Namesys.Resolve(path.Clean(p[6:]))
```
Comment (whyrusleeping) [inline core/corehttp/gateway_handler.go]: ipns can also point to a directory, so strings like: `/ipns/hash/a/b/c.avi` should be valid

File: core/corehttp/gateway_handler.go [hunk 2]
```diff
@@ -76,14 +93,14 @@ func (i *gatewayHandler) AddNodeToDAG(nd *dag.Node) (u.Key, error) {
 	return i.node.DAG.Add(nd)
 }
 
-func (i *gatewayHandler) NewDagReader(nd *dag.Node) (io.Reader, error) {
+func (i *gatewayHandler) NewDagReader(nd *dag.Node) (io.ReadSeeker, error) {
```
Comment (whyrusleeping) [inline core/corehttp/gateway_handler.go]: Also make sure to close the DagReaders! `uio.ReadSeekCloser`

File: core/corehttp/gateway.go [hunk 3]
```diff
@@ -12,5 +12,6 @@ func GatewayOption(n *core.IpfsNode, mux *http.ServeMux) error {
 		return err
 	}
 	mux.Handle("/ipfs/", gateway)
+	mux.Handle("/ipns/", gateway)
```
Comment (jbenet) [inline core/corehttp/gateway.go]: :+1:

File: core/corehttp/gateway_handler.go [hunk 4]
```diff
@@ -204,8 +220,8 @@ var listingTemplate = `
 	<h2>Index of {{ .path }}</h2>
```
Comment (whyrusleeping) [inline core/corehttp/gateway_handler.go]: *sigh* directory listings are so ugly :(

File: core/corehttp/webui.go [hunk 5]
```diff
@@ -1,25 +1,6 @@
 package corehttp
 
-import (
-	"net/http"
+// TODO: move to IPNS
```
Comment (whyrusleeping) [inline core/corehttp/webui.go]: :+1:

File: core/corehttp/gateway_handler.go (hunk 1 above)
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: btw, one thing is it doesnt have to be:

```
<ipns-hash> -> <ipfs-hash>
```

It could be

```
<ipns-hash> -> <ipfs-path>
```

So:

```
# could have mapping:
/ipns/Qmabcd -> /ipfs/Qmjklkj...../a/b/c/d/e

# so can resolve:
/ipns/Qmabcd/foo/bar -> /ipfs/Qmjklkj...../a/b/c/d/e/foo/bar
```

Comment (whyrusleeping) [inline core/corehttp/gateway_handler.go]: the pattern should be:

`/ipns/<hashA>/X -> /ipfs/<hashB>/X`

where hashB was published to hashA

File: core/corehttp/gateway_handler.go (hunk 2 above)
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: :+1:

File: core/corehttp/gateway_handler.go [hunk 6]
```diff
@@ -98,18 +115,14 @@ func (i *gatewayHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 
-	extensionIndex := strings.LastIndex(path, ".")
-	if extensionIndex != -1 {
-		extension := path[extensionIndex:]
-		mimeType := mime.TypeByExtension(extension)
-		if len(mimeType) > 0 {
-			w.Header().Add("Content-Type", mimeType)
-		}
-	}
+	w.Header().Set("X-IPFS-Path", p)
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: nice

File: core/corehttp/gateway_handler.go (hunk 1 above)
Comment (mappum) [inline core/corehttp/gateway_handler.go]: Is this the same as getting the `/ipfs/hash` that `/ipns/hash` points to, then appending `/a/b/c.avi`?

File: core/corehttp/gateway_handler.go [hunk 7]
```diff
@@ -98,18 +115,14 @@ func (i *gatewayHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 
 	dr, err := i.NewDagReader(nd)
 	if err == nil {
-		io.Copy(w, dr)
+		_, name := path.Split(urlPath)
+		// set modtime to a really long time ago, since files are immutable and should stay cached
+		modtime := time.Unix(1, 0)
+		http.ServeContent(w, r, name, modtime, dr)
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: this saddens me: 

>```
   Historically, HTTP required the Expires field-value to be no more
   than a year in the future.  While longer freshness lifetimes are no
   longer prohibited, extremely large values have been demonstrated to
   cause problems (e.g., clock overflows due to use of 32-bit integers
   for time values), and many caches will evict a response far sooner
   than that.
```
From: https://tools.ietf.org/html/rfc7234

At least we can use values larger than a year. we could set something like 5 years. that's basically forever in http browser time.

File: core/corehttp/webui.go [hunk 8]
```diff
@@ -1,25 +1,6 @@
...
+const webuiPath = "/ipfs/QmctngrQAt9fjpQUZr7Bx3BsXUcif52eZGTizWhvcShsjz"
```
Comment (jbenet) [inline core/corehttp/webui.go]: :+1:

File: cmd/ipfs/daemon.go [hunk 9]
```diff
@@ -153,9 +153,18 @@ func daemonFunc(req cmds.Request, res cmds.Response) {
...
+	}
+
 	if gatewayMaddr != nil {
 		go func() {
-			err := corehttp.ListenAndServe(node, gatewayMaddr.String(), corehttp.GatewayOption)
+			var opts = []corehttp.ServeOption{corehttp.GatewayOption}
+			if rootRedirect != nil {
+				opts = append(opts, rootRedirect)
+			}
+			err := corehttp.ListenAndServe(node, gatewayMaddr.String(), opts...)
```
Comment (jbenet) [inline cmd/ipfs/daemon.go]: this may belong in another PR, but this is the part we need to be able to start and stop on command. let's sync up with @briantigerchow on how best to do that

File: core/corehttp/gateway_handler.go (hunk 1 above)
Comment (whyrusleeping) [inline core/corehttp/gateway_handler.go]: @jbenet I dont know if i implemented that functionality... i may have to fix some things if thats the case.

**Comment by jbenet** [comment]
Know what would be _awesome_? Gateway sharness tests.

**Comment by whyrusleeping** [comment]
We could accomplish that with curl probably... would be really cool.

**Comment by whyrusleeping** [comment]
Just gotta say, im super excited to get this one merged.

File: core/corehttp/gateway_handler.go (hunk 1 above)
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: its fine if that doesnt work yet

**Comment by mappum** [comment]
Addressed your comments. :+1:

**Comment by whyrusleeping** [comment]
@jbenet we dont have the path types yet, thats on my todo list.

**Comment by jbenet** [comment]
> @jbenet we dont have the path types yet, thats on my todo list.

i just mean string constants

**Comment by whyrusleeping** [comment]
Im okay with sharness tests being added in another PR, as long as it happens pretty soon. @jbenet what do you think?

**Comment by jbenet** [comment]
> Im okay with sharness tests being added in another PR, as long as it happens pretty soon. @jbenet what do you think?

thoughts mappum? sharness tests tend to not get done if not with the features.

**Comment by mappum** [comment]
I'll make some basic tests, we can just add more advanced ones when we find bugs we don't want to regress.

**Comment by whyrusleeping** [comment]
And of course, the tests are all green for your PR, but on mine, they will randomly fail on unrelated issues :(

**Comment by mappum** [comment]
>And of course, the tests are all green for your PR, but on mine, they will randomly fail on unrelated issues :(

That's usually what happens for me, this last commit is just lucky. :four_leaf_clover:

**Comment by jbenet** [comment]
let's get rid of all the timing / broken tests

File: test/sharness/t0100-gateway.sh [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+'
+
+test_expect_success "GET IPFS directory file output looks good" '
+  test_cmp dir/test actual
+'
+
+test_expect_failure "GET IPNS path succeeds" '
+  ipfs name publish "$HASH" &&
+  NAME=`ipfs config Identity.PeerID` &&
+  wget http://127.0.0.1:5001/ipns/"$NAME" -O actual
```
Comment (jbenet) [inline test/sharness/t0100-gateway.sh]: here want to escape like this:
```
wget "http://127.0.0.1:5001/ipns/$NAME" -O actual
```

**Comment by jbenet** [comment]
the tar error is still here:

```
gzip: stdin: unexpected end of file
tar: Child returned status 1
tar: Error is not recoverable: exiting now
not ok 22 - gzipped tar archive output is valid (directory)
#	
#	  tar -zxf "$HASH2".tar.gz &&
#	  test_cmp dir/a "$HASH2"/a &&
#	  test_cmp dir/b/c "$HASH2"/b/c &&
#	  rm -r "$HASH2"
#	

```

**Comment by mappum** [comment]
>the tar error is still here

It only fails on Jenkins, could it be an incompatible version of gzip?

**Comment by jbenet** [comment]
yeah maybe.

**Comment by jbenet** [comment]
i just rebased the last bits. after this runs should be RFM

**Comment by whyrusleeping** [comment]
We now have two t0100-* tests.

**Comment by jbenet** [comment]
good catch
